server/driver/wivrn_session.cpp

```diff
@@ -484,8 +515,21 @@ void wivrn_session::operator()(from_headset::inputs && inputs)
 {
 	auto offset = get_offset();
 
-	left_hand.set_inputs(inputs, offset);
-	right_hand.set_inputs(inputs, offset);
+	if (roles.left_profile != XRT_DEVICE_INVALID)
+	{
+		if (roles.left == left_hand_interaction_index)
+			left_hand_interaction.set_inputs(inputs, offset);
+		else
+			left_controller.set_inputs(inputs, offset);
+	}
+
+	if (roles.right_profile != XRT_DEVICE_INVALID)
+	{
+		if (roles.right == right_hand_interaction_index)
+			right_hand_interaction.set_inputs(inputs, offset);
+		else
+			right_controller.set_inputs(inputs, offset);
+	}
```

Comment:
Just check `roles.left`, with an if/else if. The `roles.left_profile` should not be relevant here.